Src/Sarif.PatternMatcher/RegexMetadata.cs

```diff
@@ -0,0 +1,30 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+
+namespace Microsoft.CodeAnalysis.Sarif.PatternMatcher
+{
+    [Flags]
+    public enum RegexMetadata
+    {
+        /// <summary>
+        /// Implicitly, the absence of any regex metadata means
+        /// that it is a required match. That is, if the regex
+        /// does not exist, the analysis rules halts immediately.
+        /// </summary>
+        None = 0,
+
+        /// <summary>
+        /// Indicates that the regex retrieves an optionally
+        /// available value.
+        /// </summary>
+        Optional = 0x1,
+
+        /// <summary>
+        /// Indicates that the regex is potentially shared across
+        /// multiple rules (and therefore can be cached).
+        /// </summary>
+        Shared = 0x2,
```

Comment:
>Shared [](http://example.com/codeflow?start=8&length=6)

Unused for now, but there is an intended future use we have in mind. Another interesting case is to mark a regex as one that, whenever it matches, should fire a warning but only if another rule hasn't already interacted with it. 

So, for example, many rules may consume a 256 bit symmetric key. If any rule seems to think a key of this kind 'belongs' to it, great. If no rule does, the scanner might emit a warning which basically notes 'this file has something that looks like a secret but no rule recognized it as relevant to it.'